Ch4 Q13)

This question should be answered using the Weekly data set, which
is part of the ISLP package. This data is similar in nature to the
Smarket data from this chapter’s lab, except that it contains 1, 089
weekly returns for 21 years, from the beginning of 1990 to the end of
2010.

(e) Repeat (d) using LDA

In [2]:
pip install ISLP

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 48.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.3/349.3 kB 26.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 522.0/522.0 kB 32.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.4/36.4 MB 44.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 815.2/815.2 kB 40.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 890.6/890.6 kB 46.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.5/94.5 kB 7.7 MB/s eta 0:00:00
  Created wheel for autograd-gamma: filename=autograd_gamma-0.5.0-py3-none-any.whl size=4031 sha256=f52f03a90d226dbd6eae85ae871e0639c1e4251163e2513ffc60074e573944ef
  Stored in directory: /root/.cache/pip/wheels/25/cc/e0/ef2969164144c899fedb22b338f6703e2b9cf46eeebf254991
Successfully built autograd-gamma
  Attempting un

In [3]:
import pandas as pd
from ISLP import load_data
from ISLP import confusion_table
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score

# Load the "Weekly" dataset
weekly = load_data('Weekly')

# Create a binary response variable (1 for "Up," 0 for "Down")
weekly['Direction_binary'] = (weekly['Direction'] == 'Up').astype(int)

# Define the training and testing data periods
train_data = weekly[weekly['Year'] <= 2008]
test_data = weekly[weekly['Year'] >= 2009]

# Select the "Lag2" variable as the predictor for LDA
X_train = train_data[['Lag2']]
y_train = train_data['Direction_binary']

X_test = test_data[['Lag2']]
y_test = test_data['Direction_binary']

# Fit the LDA model using only "Lag2" as the predictor
model = LinearDiscriminantAnalysis()
model.fit(X_train, y_train)

# Predict using the test data
lda_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(lda_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, lda_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0           9   5
1          34  56
Overall Fraction of Correct Predictions: 0.6250


(f) Repeat (d) using QDA.

In [4]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

# Fit the QDA model using only "Lag2" as the predictor
model = QuadraticDiscriminantAnalysis()
model.fit(X_train, y_train)

# Predict using the test data
qda_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(qda_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, qda_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0           0   0
1          43  61
Overall Fraction of Correct Predictions: 0.5865


(g) Repeat (d) using KNN with K = 1.

In [5]:
from sklearn.neighbors import KNeighborsClassifier

# Fit the KNN model with K=1 using only "Lag2" as the predictor
model = KNeighborsClassifier(n_neighbors=1)
model.fit(X_train, y_train)

# Predict using the test data
knn_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(knn_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, knn_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          22  32
1          21  29
Overall Fraction of Correct Predictions: 0.4904


(h) Repeat (d) using naive Bayes.

In [6]:
from sklearn.naive_bayes import GaussianNB

# Fit the Naive Bayes model using Gaussian Naive Bayes with only "Lag2" as the predictor
model = GaussianNB()
model.fit(X_train, y_train)

# Predict using the test data
gnb_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(gnb_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, gnb_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0           0   0
1          43  61
Overall Fraction of Correct Predictions: 0.5865


(i) Which of these methods appears to provide the best results on this data?

To answer the question of which method provide the best results on the data, we can use the accuracy as a standard. According to the calculated accuracy on each method, we can said that the Linear Discriminant Analysis method is providing the best results on the data, as it have a highest 0.6250 accuracy.

(j) Experiment with different combinations of predictors, including
possible transformations and interactions, for each of the
methods. Report the variables, method, and associated confusion
matrix that appears to provide the best results on the held
out data. Note that you should also experiment with values for
K in the KNN classifier.

Load data and do train-test split

In [8]:
import pandas as pd
from ISLP import load_data
from ISLP import confusion_table
from sklearn.metrics import accuracy_score

# Load the "Weekly" dataset
weekly = load_data('Weekly')

# Create a binary response variable (1 for "Up," 0 for "Down")
weekly['Direction_binary'] = (weekly['Direction'] == 'Up').astype(int)

# Define the training and testing data periods
train_data = weekly[weekly['Year'] <= 2008]
test_data = weekly[weekly['Year'] >= 2009]

# Select the "Lag2" variable as the predictor for LDA
X_train = train_data[['Lag2']]
y_train = train_data['Direction_binary']

X_test = test_data[['Lag2']]
y_test = test_data['Direction_binary']

Logistic Regression:

Origin accuracy 0.5481

In [10]:
import statsmodels.api as sm

# Define the training and testing data periods
train_data = weekly[weekly['Year'] <= 2008]
test_data = weekly[weekly['Year'] >= 2009]

# Select the "Lag2" variable as the predictor for the model
X_train = train_data[['Lag2']]
y_train = train_data['Direction_binary']

X_test = test_data[['Lag2']]
y_test = test_data['Direction_binary']

# Add a constant to the predictors (intercept)
x_train = sm.add_constant(X_train)

# Fit the logistic regression model
model = sm.Logit(y_train, X_train)
result = model.fit()

# Print the summary of the logistic regression results
print(result.summary())

# Predict using the test data
y_pred = result.predict(X_test)>0.5 # Set a threshold of 0.5 for binary classification

# Compute the confusion matrix
confusion = confusion_table(y_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, y_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Optimization terminated successfully.
         Current function value: 0.690654
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:       Direction_binary   No. Observations:                  985
Model:                          Logit   Df Residuals:                      984
Method:                           MLE   Df Model:                            0
Date:                Sun, 03 Nov 2024   Pseudo R-squ.:               -0.004340
Time:                        03:17:48   Log-Likelihood:                -680.29
converged:                       True   LL-Null:                       -677.35
Covariance Type:            nonrobust   LLR p-value:                       nan
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Lag2           0.0629      0.029      2.192      0.028       0.007       0.119
Confusion Matrix:
Truth      

Add Today as predictors: New accuracy 1.0000

In [11]:
from sklearn.linear_model import LogisticRegression

# Add Today as predictors
X_train = train_data[['Today', 'Lag2']]
X_test = test_data[['Today', 'Lag2']]

# Fit the logistic regression model with the new predictors
model = LogisticRegression()
model.fit(X_train, y_train)

# Predict using the test data
y_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(y_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, y_pred)

# Print results
print("Logistic Regression with Interaction Term - Results:")
print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Logistic Regression with Interaction Term - Results:
Confusion Matrix:
Truth       0   1
Predicted        
0          43   0
1           0  61
Overall Fraction of Correct Predictions: 1.0000


Add Today and the interaction term (Lag1 * Lag2) as predictors: New accuracy 1.0000

In [12]:
from sklearn.linear_model import LogisticRegression

# Add Today and the interaction term (Lag1 * Lag2) as predictors
X_train = train_data[['Today', 'Lag2']]
X_train['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']

X_test = test_data[['Today', 'Lag2']]
X_test['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']

# Fit the logistic regression model with the new predictors
model = LogisticRegression()
model.fit(X_train, y_train)

# Predict using the test data
y_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(y_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, y_pred)

# Print results
print("Logistic Regression with Interaction Term - Results:")
print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Logistic Regression with Interaction Term - Results:
Confusion Matrix:
Truth       0   1
Predicted        
0          43   0
1           0  61
Overall Fraction of Correct Predictions: 1.0000


<ipython-input-12-cf52c6c764cc>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']
<ipython-input-12-cf52c6c764cc>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']


Linear Discriminant Analysis (LDA):

Original LDA(part e) accuracy: 0.6250

Add Today as predictors: 0.9615


In [13]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

# Add Today as predictors
X_train_lda = train_data[['Today', 'Lag2']]
X_test_lda = test_data[['Today', 'Lag2']]

# Fit the LDA model with the new predictors
model = LinearDiscriminantAnalysis()
model.fit(X_train_lda, y_train)

# Predict using the test data
lda_pred = model.predict(X_test_lda)

# Compute the confusion matrix
confusion = confusion_table(lda_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, lda_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          39   0
1           4  61
Overall Fraction of Correct Predictions: 0.9615


Add Today and the interaction term (Lag1 * Lag2) as predictors: 0.9615

In [14]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

# Add Today and the interaction term (Lag1 * Lag2) as predictors
X_train_lda = train_data[['Today', 'Lag2']]
X_train_lda['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']

X_test_lda = test_data[['Today', 'Lag2']]
X_test_lda['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']


# Fit the LDA model with the new predictors
model = LinearDiscriminantAnalysis()
model.fit(X_train_lda, y_train)

# Predict using the test data
lda_pred = model.predict(X_test_lda)

# Compute the confusion matrix
confusion = confusion_table(lda_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, lda_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          39   0
1           4  61
Overall Fraction of Correct Predictions: 0.9615


<ipython-input-14-cb407402f24d>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train_lda['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']
<ipython-input-14-cb407402f24d>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test_lda['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']


Quadratic Discriminant Analysis (QDA):

The Original QDA(part f) accuracy is 0.5865, after trying to add Today or both Today and the interaction term (Lag1 * Lag2) as predictors, both of their accuracy increases. Adding Today improve the accuracy to 0.9808, and if also adding the lag1_lag2 interaction term, the accuracy decrease to 0.9615.

After add Today as predictors, accuracy change to: 0.9808

In [15]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

# Try using Today and Lag2 as predictors
X_train_qda = train_data[['Today', 'Lag2']]
X_test_qda = test_data[['Today', 'Lag2']]

# Fit the LDA model with the new predictors
model = QuadraticDiscriminantAnalysis()
model.fit(X_train_qda, y_train)

# Predict using the test data
qda_pred = model.predict(X_test_qda)

# Compute the confusion matrix
confusion = confusion_table(qda_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, qda_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          41   0
1           2  61
Overall Fraction of Correct Predictions: 0.9808


Add Today and the interaction term (Lag1 * Lag2) as predictors: 0.9615

In [16]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

# Add Today and the interaction term (Lag1 * Lag2) as predictors
X_train_qda = train_data[['Today', 'Lag2']]
X_train_qda['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']

X_test_qda = test_data[['Today', 'Lag2']]
X_test_qda['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']

# Fit the LDA model with the new predictors
model = QuadraticDiscriminantAnalysis()
model.fit(X_train_qda, y_train)

# Predict using the test data
qda_pred = model.predict(X_test_qda)

# Compute the confusion matrix
confusion = confusion_table(qda_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, qda_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          41   2
1           2  59
Overall Fraction of Correct Predictions: 0.9615


<ipython-input-16-084f95247d06>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train_qda['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']
<ipython-input-16-084f95247d06>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test_qda['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']


K-Nearest Neighbors (KNN):

The accuracy of k=1 is 0.4904 according to part g. After testing for k=3,5,7, when k=3 it provide highest accuracy of 0.9135.

In [19]:
from sklearn.neighbors import KNeighborsClassifier

# Fit the KNN model with K=5 using only "Lag2" as the predictor
model = KNeighborsClassifier(n_neighbors=5)
model.fit(X_train, y_train)

# Predict using the test data
knn_pred = model.predict(X_test)

# Compute the confusion matrix
confusion = confusion_table(knn_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, knn_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          37   3
1           6  58
Overall Fraction of Correct Predictions: 0.9135


Naive Bayes:

The original accuracy in part h is 0.5865, after trying to add Today or both Today and the interaction term (Lag1 * Lag2) as predictors, both of their accuracy increases. Adding Today as predictors, the accuracy change to 0.9615, and if also adding the lag1_lag2 interaction term, the accracy further improve to 0.9904.


In [20]:
from sklearn.naive_bayes import GaussianNB

# Try using Today and Lag2 as predictors
X_train_gnb = train_data[['Today', 'Lag2']]
X_test_gnb = test_data[['Today', 'Lag2']]

# Fit the Naive Bayes model using Gaussian Naive Bayes with only "Lag2" as the predictor
model = GaussianNB()
model.fit(X_train_gnb, y_train)

# Predict using the test data
gnb_pred = model.predict(X_test_gnb)

# Compute the confusion matrix
confusion = confusion_table(gnb_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, gnb_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          39   0
1           4  61
Overall Fraction of Correct Predictions: 0.9615


In [21]:
from sklearn.naive_bayes import GaussianNB

# Add Today and the interaction term (Lag1 * Lag2) as predictors
X_train_gnb = train_data[['Today', 'Lag2']]
X_train_gnb['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']

X_test_gnb = test_data[['Today', 'Lag2']]
X_test_gnb['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']
# Fit the Naive Bayes model using Gaussian Naive Bayes with only "Lag2" as the predictor
model = GaussianNB()
model.fit(X_train_gnb, y_train)

# Predict using the test data
gnb_pred = model.predict(X_test_gnb)

# Compute the confusion matrix
confusion = confusion_table(gnb_pred, y_test)

# Compute the overall fraction of correct predictions (accuracy)
accuracy = accuracy_score(y_test, gnb_pred)

print("Confusion Matrix:")
print(confusion)
print(f"Overall Fraction of Correct Predictions: {accuracy:.4f}")

Confusion Matrix:
Truth       0   1
Predicted        
0          42   0
1           1  61
Overall Fraction of Correct Predictions: 0.9904


<ipython-input-21-d3afedc95ba4>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train_gnb['Lag1_Lag2_interaction'] = train_data['Lag1'] * train_data['Lag2']
<ipython-input-21-d3afedc95ba4>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test_gnb['Lag1_Lag2_interaction'] = test_data['Lag1'] * test_data['Lag2']


Conclusion:

After experimenting with different methods and according to the result, whether add Today or both today and the interaction term (Lag1 * Lag2) as predictors, it increases the accuracy of the model to 1.0000. In this case, as both combinations of predictors can max out the accuracy, to judge which appears to provide the best results on the held out data, we can consider the model's simplicity. A more simple but more accurate model is better, therefore the logistic regression model with predictors Today and lag2 appear to provide the best results on the held out data under the overall consideration.